### PR TITLE
@task PDA-1410 Create bi-direactional method updateItemsInStock

### PR DESCRIPTION
**Pull request by dtang-parasoft**

### CODE REVIEW

**Comment by yshi-parasoft** [comment]
Please merge conflict.
Also I just realized that we need to add log for exceptions. Please add log for this method and also help to add log for the other two existing methods for the JSON service.

**Comment by yshi-parasoft** [comment]
I think we should handle the error case when the stock is already 0 and we performed a Deduction on it. 
Also did you test multi-thread situation? I didn't see any special handling for it and if not please research for possible issues.

File: src/main/java/com/parasoft/demoapp/messages/AssetMessages.java [hunk 1]
```diff
@@ -29,12 +29,14 @@ public class AssetMessages extends Messages {
...
 			"There is not enough inventory to fulfill your order. Edit your requisition and resubmit.";
-	public static final String IN_STOCK_OF_CART_ITEM_IS_INSUFFICIENT = "Inventory is not enough.";
+	public static final String INVENTORY_IS_NOT_ENOUGH = "Inventory is not enough.";
 	public static final String THIS_ITEM_IS_NOT_IN_THE_SHOPPING_CART = "This item is not in the shopping cart.";
 	public static final String NO_CART_ITEMS = "No cart items.";
 	public static final String THERE_IS_NO_CART_ITEM_CORRESPONDING_TO = "There is no cart item corresponding to {0}.";
 	public static final String SEARCH_FIELD_CANNOT_BE_BLANK = "Search field cannot be an empty string(null, '' or '  ').";
 	public static final String INVENTORY_NOT_FOUND_WITH_ITEM_ID = "Inventory not found with item id {0}.";
+	public static final String INCORRECT_OPERATION = "Incorrect operation";
+	public static final String REQUEST_PARAMETER_CANNOT_BE_EMPTY="The request parameter cannot be empty.";
```
Comment (rwang-parasoft) [inline src/main/java/com/parasoft/demoapp/messages/AssetMessages.java:39]: Change it to REQUEST_PARAMETER_CANNOT_BE_NULL and need to change the value.

File: src/main/java/com/parasoft/demoapp/grpc/service/JsonServiceImpl.java [hunk 2]
```diff
@@ -81,4 +94,68 @@ public void getItemsInStock(StreamObserver<ItemEntity> responseObserver) {
             log.error(e.getMessage(), e);
         }
     }
+
+    @Override
+    public StreamObserver<ItemRequest> updateItemsInStock(StreamObserver<ItemResponse> responseObserver) {
+        return new StreamObserver<ItemRequest>() {
+            @Override
+            public void onNext(ItemRequest value) {
```
Comment (rwang-parasoft) [inline src/main/java/com/parasoft/demoapp/grpc/service/JsonServiceImpl.java:102]: Rename the vaue itemRequest.

File: src/main/java/com/parasoft/demoapp/grpc/service/JsonServiceImpl.java [hunk 3]
```diff
@@ -81,4 +94,68 @@ public void getItemsInStock(StreamObserver<ItemEntity> responseObserver) {
...
+                try {
+                    synchronized (clock) {
+                        ParameterValidator.requireNonNull(value, AssetMessages.REQUEST_PARAMETER_CANNOT_BE_EMPTY);
+                        ParameterValidator.requireNonNull(value.getValue(), AssetMessages.IN_STOCK_CANNOT_BE_NULL);
```
Comment (rwang-parasoft) [inline src/main/java/com/parasoft/demoapp/grpc/service/JsonServiceImpl.java:106]: Change the message to OPERATION_NUMBER_CANNOT_BE_NULL=The quantity for item inventory dedction or addition can not be null.

File: src/main/java/com/parasoft/demoapp/grpc/service/JsonServiceImpl.java [hunk 4]
```diff
@@ -81,4 +94,68 @@ public void getItemsInStock(StreamObserver<ItemEntity> responseObserver) {
...
+                        Integer newInStock = 0;
+                        ItemEntity item = itemService.getItemById(value.getId());
+                        Integer inStock = item.getInStock();
+
+                        if (value.getOperation() == null) {
+                            IncorrectOperationException incorrectOperationException =
```
Comment (rwang-parasoft) [inline src/main/java/com/parasoft/demoapp/grpc/service/JsonServiceImpl.java:112]: Do not create a new IncorrectOperationException.

**Review by rwang-parasoft — APPROVED** [review]

**Comment by yshi-parasoft** [comment]
The changes looks good. Please resolve conflict before merging.

**Review by yshi-parasoft — CHANGES_REQUESTED** [review]
The changes looks good. Except that I think we can change the object name to be "lock" instead of "clock". Also please resolve conflict before merging.

**Review by yshi-parasoft — APPROVED** [review]

**Pull request merged**